File: backend/app/schemas/missions_admin.py
"""Mission authoring schemas (P5-4) — `/missions/admin/*`.

Variant `config` validation for the `quiz` kind reuses `AdminContentQuiz`
directly (`schemas/lms_admin.py`) rather than a second copy of the same
shape/validator (exactly-one-correct-option, B3) — one place a quiz
question can be malformed, whichever surface authored it.
"""

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field

MissionKind = Literal["design", "submission", "quiz", "checklist", "operate", "external"]
MissionTeamPolicy = Literal["solo", "team", "either"]
MissionStatus = Literal["draft", "in_review", "published", "archived"]
MissionAccessMode = Literal["open", "invite"]  # never 'paid' — see models/missions/mission.py


class MissionCreate(BaseModel):
    title: str
    slug: str
    summary: str | None = None
    description: str | None = None
    kind: MissionKind
    team_policy: MissionTeamPolicy = "solo"
    access_mode: MissionAccessMode = "open"
    track: str | None = None


class MissionUpdate(BaseModel):
    title: str | None = None
    summary: str | None = None
    description: str | None = None
    status: MissionStatus | None = None
    access_mode: MissionAccessMode | None = None
    team_policy: MissionTeamPolicy | None = None
    track: str | None = None


class MissionVariantCreate(BaseModel):
    label: str
    position: int
    points: int = Field(ge=0)
    config: dict = {}


class MissionVariantUpdate(BaseModel):
    label: str | None = None
    position: int | None = None
    points: int | None = Field(default=None, ge=0)
    config: dict | None = None


class MissionVariantAdminOut(BaseModel):
    id: UUID
    label: str
    position: int
    points: int
    config: dict


class MissionAdminOut(BaseModel):
    id: UUID
    title: str
    slug: str
    summary: str | None = None
    description: str | None = None
    kind: str
    team_policy: str
    status: str
    access_mode: str
    track: str | None = None
    image_url: str | None = None
    authored_by: UUID
    authored_by_name: str | None = None
    reviewed_by: UUID | None = None
    created_at: datetime | None = None
    variants: list[MissionVariantAdminOut] = []


class MissionAttemptReviewIn(BaseModel):
    passed: bool
    score: float | None = None
    review_comment: str | None = None


class MissionAttemptAssignIn(BaseModel):
    """The general ops-assign shape (2026-08-21, LMS Program redesign) —
    the only way a solo attempt gets a `cohort_id` now that self-started
    attempts are always independent. `variant_id=None` resolves to the
    mission's easiest variant. `force_new=True` only for re-assigning a
    student who already passed a prior run."""
    user_id: UUID
    mission_id: UUID
    cohort_id: UUID
    variant_id: UUID | None = None
    force_new: bool = False


class MissionAttemptAdminOut(BaseModel):
    id: UUID
    mission_id: UUID
    mission_title: str
    variant_id: UUID
    variant_label: str
    # user_id XOR team_id — solo attempts set the former, team attempts (P6-2)
    # the latter, mirroring mission_attempts' own CHECK constraint.
    user_id: UUID | None = None
    student_name: str | None = None
    team_id: UUID | None = None
    team_name: str | None = None
    attempt_no: int
    status: str
    score: float | None = None
    payload: dict = {}
    started_at: datetime | None = None
    submitted_at: datetime | None = None
    decided_at: datetime | None = None


class MissionTeamCreateAdminIn(BaseModel):
    """Ops-assign (P6-4): cohort_id required — self-form (schemas/teams.py
    ::TeamCreateIn) is the same primitive without one."""
    name: str
    cohort_id: UUID
    member_ids: list[UUID] = []


class MissionTeamAdminOut(BaseModel):
    id: UUID
    name: str
    cohort_id: UUID | None = None
    member_ids: list[UUID] = []
    member_names: list[str] = []




class MissionAssignmentOut(BaseModel):
    id: UUID
    user_id: UUID
    user_name: str
    user_email: str
    mission_id: UUID
    source: str
    status: str
    granted_by: UUID | None = None
    created_at: datetime | None = None


class MissionAssignmentGrantIn(BaseModel):
    user_id: UUID


class MissionBulkAssignIn(BaseModel):
    """Role-only — missions have no cohort curriculum table the way courses
    do, so there is no cohort-based bulk path here (BulkGrantIn's other
    branch)."""
    role: str


class MissionBulkAssignOut(BaseModel):
    granted: int
    already_assigned: int

File: frontend/src/api/missions_admin.ts
/** Missions admin API (P5-4) — thin wrapper over `/missions/admin`. */
import { api } from "@/api/client";
import type { MissionTeamPolicy } from "@/api/missions";

export type { MissionTeamPolicy };

export interface MissionAdminOption {
  id: string;
  title: string;
}

export const listMissionsAdminApi = () =>
  api.get<MissionAdminOption[]>("/missions/admin").then((r) => r.data);

export type MissionKind = "design" | "submission" | "quiz" | "checklist" | "operate" | "external";
export type MissionStatus = "draft" | "in_review" | "published" | "archived";
export type MissionAccessMode = "open" | "invite";

export interface MissionVariantAdmin {
  id: string;
  label: string;
  position: number;
  points: number;
  config: Record<string, unknown>;
}

export interface MissionAdmin {
  id: string;
  title: string;
  slug: string;
  summary: string | null;
  description: string | null;
  kind: MissionKind;
  team_policy: MissionTeamPolicy;
  status: MissionStatus;
  access_mode: MissionAccessMode;
  track: string | null;
  image_url: string | null;
  authored_by: string;
  authored_by_name: string | null;
  reviewed_by: string | null;
  created_at: string | null;
  variants: MissionVariantAdmin[];
}

export interface MissionUpdateIn {
  status?: MissionStatus;
  title?: string;
  summary?: string;
  description?: string;
  access_mode?: MissionAccessMode;
  team_policy?: MissionTeamPolicy;
  track?: string;
}

/** Full rows (status, kind, variants, ...) for the ops mission list —
 * same endpoint `listMissionsAdminApi` uses, kept separate so the
 * prerequisites picker's narrow type doesn't have to change shape. */
export const listMissionsAdminFullApi = () =>
  api.get<MissionAdmin[]>("/missions/admin").then((r) => r.data);

export const getMissionAdminApi = (id: string) =>
  api.get<MissionAdmin>(`/missions/admin/${id}`).then((r) => r.data);

export const updateMissionAdminApi = (id: string, data: MissionUpdateIn) =>
  api.patch<MissionAdmin>(`/missions/admin/${id}`, data).then((r) => r.data);

export const deleteMissionAdminApi = (id: string) =>
  api.delete<void>(`/missions/admin/${id}`).then((r) => r.data);

// ── mission assignment (2026-08-12) ─────────────────────────────────────────

export interface MissionAssignment {
  id: string;
  user_id: string;
  user_name: string;
  user_email: string;
  mission_id: string;
  source: string;
  status: string;
  granted_by: string | null;
  created_at: string | null;
}

export interface MissionBulkAssignResult {
  granted: number;
  already_assigned: number;
}

export const listMissionRosterApi = (missionId: string) =>
  api.get<MissionAssignment[]>(`/missions/admin/${missionId}/roster`).then((r) => r.data);

export const grantMissionAssignmentApi = (missionId: string, userId: string) =>
  api.post<MissionAssignment>(`/missions/admin/${missionId}/assignments`, { user_id: userId }).then((r) => r.data);

export const bulkGrantMissionAssignmentApi = (missionId: string, role: string) =>
  api.post<MissionBulkAssignResult>(`/missions/admin/${missionId}/assignments/bulk`, { role }).then((r) => r.data);

export const revokeMissionAssignmentApi = (assignmentId: string) =>
  api.post<MissionAssignment>(`/missions/admin/assignments/${assignmentId}/revoke`).then((r) => r.data);

// ── mission authors (mission_managers, 7B-7) ─────────────────────────────
//
// The many-to-many table and its endpoints have existed since 7B-7 but were
// never wired to any screen, so a mission's authorship could only be changed
// by writing to the database. Ops assigns; an assigned author gets the
// mission's stats, its review queue and its teaching content — not its
// grading thresholds, which stay frozen while published.

export interface MissionAuthor {
  user_id: string;
  full_name: string;
  granted_by: string | null;
  created_at: string | null;
}

export const listMissionAuthorsApi = (missionId: string) =>
  api.get<MissionAuthor[]>(`/missions/admin/${missionId}/managers`).then((r) => r.data);

export const addMissionAuthorApi = (missionId: string, userId: string) =>
  api.post<MissionAuthor>(`/missions/admin/${missionId}/managers`, { user_id: userId }).then((r) => r.data);

export const removeMissionAuthorApi = (missionId: string, userId: string) =>
  api.delete(`/missions/admin/${missionId}/managers/${userId}`).then((r) => r.data);
